src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRun/GCRefMapNode.cs

```diff
@@ -0,0 +1,121 @@
+            _methods = new List<IMethodNode>();
+            _index = 0;
+        }
+
+        public override ObjectNodeSection Section => ObjectNodeSection.ReadOnlyDataSection;
+
+        public override bool IsShareable => false;
+
+        public override int ClassCode => 555444333;
+
+        public override bool StaticDependenciesAreComputed => true;
+
+        public int Offset => 0;
+
+        public void AddImport(Import import)
+        {
+            if (import is IMethodNode methodNode)
+            {
+                while (_methods.Count <= _index)
+                {
+                    _methods.Add(null);
+                }
+                _methods[_index] = methodNode;
+            }
+            _index++;
+        }
+
+        public void AppendMangledName(NameMangler nameMangler, Utf8StringBuilder sb)
+        {
+            sb.Append("GCRefMap->");
+            sb.Append(_importSection.Name);
+        }
+
+        public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
+        {
+            if (_methods.Count == 0 || relocsOnly)
+            {
+                return new ObjectData(
+                    data: Array.Empty<byte>(), 
+                    relocs: Array.Empty<Relocation>(), 
+                    alignment: 1, 
+                    definedSymbols: new ISymbolDefinitionNode[] { this });
+            }
+
+            GCRefMapBuilder builder = new GCRefMapBuilder(factory, relocsOnly);
+            builder.Builder.RequireInitialAlignment(4);
+            builder.Builder.AddSymbol(this);
+
+            // First, emit the initial ref map offset and reserve the offset map entries
+            int offsetCount = _methods.Count / GCREFMAP_LOOKUP_STRIDE;
+            builder.Builder.EmitInt((offsetCount + 1) * sizeof(int));
+
+            ObjectDataBuilder.Reservation[] offsets = new ObjectDataBuilder.Reservation[offsetCount];
+            for (int offsetIndex = 0; offsetIndex < offsetCount; offsetIndex++)
+            {
+                offsets[offsetIndex] = builder.Builder.ReserveInt();
+            }
+
+            // Next, generate the actual method GC ref maps and update the offset map
+            int nextOffsetIndex = 0;
+            int nextMethodIndex = GCREFMAP_LOOKUP_STRIDE - 1;
+            for (int methodIndex = 0; methodIndex < _methods.Count; methodIndex++)
+            {
+                if (methodIndex >= nextMethodIndex)
+                {
+                    builder.Builder.EmitInt(offsets[nextOffsetIndex], builder.Builder.CountBytes);
+                    nextOffsetIndex++;
+                    nextMethodIndex += GCREFMAP_LOOKUP_STRIDE;
+                }
+                IMethodNode methodNode = _methods[methodIndex];
+                if (methodNode is LocalMethodImport localMethodImport && localMethodImport.MethodCodeNode.IsEmpty)
+                {
+                    // Emit an empty entry into the GC ref map for the uncompilable method
```

Comment:
Is this correct? I would think we need the GCRef even for methods that are not compiled into the image.

Reply:
Well, I think the purpose of this conditional block is kind of the other way around - this is to cater for uncompilable methods (e.g. because they reference S.N.Vector) by emitting an empty GC ref map just so that the method import cell array and the GC ref map array don't fall out of sync.

Reply:
I think we need the GC ref map even for methods that we were not able to compile. Otherwise, we won't be able to do GC stack scan when those methods are called at runtime and we fallback to JIT.

Reply:
Hmm, how can we encode GC ref map for a function with a Vector parameter if Vector has runtime-determined size?

Reply:
There are two options how to deal with variable sizes Vector:
- Do not compile any methods whose code depends on the variable-sized Vector. It is what crossgen does today.
- Compile methods whose depend on Vector size and emit the `READYTORUN_FIXUP_Check_TypeLayout` fixup that will encode the Vector size assumed by the method code. This fixup will cause the method code to be rejected at runtime if the runtime Vector size is different from the Vector size assumed by the method code.
   - Advanced version of this is the code multi-versioning that we keep talking about every once in a while.

Reply:
Thanks for the clarification, this makes sense to me. You have also nicely ordered the three options by their complexity. In this stage I'm heads down pulling off at least option (1); once I add support for type layout checks, I can implement (2) and we can start a technical discussion about (3).

I assume that this most complex option will likely require some additions to the R2R format as I suppose we should end up with a map of variant entrypoints keyed by some processor extension identifier. I don't yet have a full picture of all the extensions / coprocessor / instruction set extensions that are available in practice and supported by JIT; whether they are all mutually exclusive or whether there are some "orthogonal" sets of extensions that may or may not be optionally present on a target machine (so that we'd need a bitmask for the extension ID).

Reply:
In fact I'm just an idiot, I just had a blind spot and you were right in the very first place - when a method fails compiling (e.g. because it uses runtime JIT), we still need to be able to emit GC ref map for calls to it. I have hit this in one of the Top200 tests after I started running CPAOT tests with GC stress per your suggestion. I already have a local fix I'm testing right now.